### PR TITLE
EDM-240: ensure application status

### PR DESCRIPTION
**Pull request by hexfusion**
This PR follows up on [#561](https://github.com/flightctl/flightctl/pull/561) by strengthening the application flow and enhancing status reporting. It introduces a priority queue that optimizes reconciliation logic by implementing backoff durations for resyncing template versions, thereby reducing resource overhead in the study state. Additionally, the priority queue enables a more granular approach by allowing the setting of maximum retries and managing failed versions, resulting in improved control and stability of the application.

For example, we can now poll for updates every 10s and then only resync the current version every 5m. This flow ensures our on-disk and in-memory state. While also instantly starting the reconciliation of new template versions when observed.

This PR also adds a centralized error package, which is the groundwork for human readable status messages. The full status message update will be in the next PR.  The initial implementation adds support for declaring retryable vs non-retryable errors.

TODO:
- [x] add timed delay requeue of current rendered status vs constant resync
- [x] cleanup logging
- [x] fix spec tests
- [ ] update status messages to only use errors.ErrFoo and log rest (follow-up)

### CODE REVIEW

File: internal/agent/device/bootstrap.go [hunk 1]
```diff
@@ -213,31 +213,7 @@ func (b *Bootstrap) ensureBootedOS(ctx context.Context, desired *v1alpha1.Render
...
-		}),
-		status.SetConfig(v1alpha1.DeviceConfigStatus{
-			RenderedVersion: desired.RenderedVersion,
-		}),
-	}
-
-	_, updateErr := b.statusManager.Update(ctx, updateFns...)
-	if updateErr != nil {
-		b.log.Warnf("Failed setting status: %v", updateErr)
-	}
```
Comment (hexfusion) [inline internal/agent/device/bootstrap.go:240]: this wasn't needed in bootstap reconciliation should complete on next sync during upgrade.

File: internal/agent/device/spec/spec.go [hunk 2]
```diff
@@ -84,6 +77,12 @@ type SpecManager struct {
 	managementClient client.Management
 	bootcClient      container.BootcClient
 
+	// cached rendered versions
+	currentRenderedVersion  string
+	desiredRenderedVersion  string
+	rollbackRenderedVersion string
```
Comment (hexfusion) [inline internal/agent/device/spec/spec.go:83]: caching this locally eliminates the need to pull the spec into memory to check the version.

File: internal/agent/device/spec/manager.go [hunk 3]
```diff
@@ -0,0 +1,524 @@
...
+	"github.com/flightctl/flightctl/internal/agent/client"
+	"github.com/flightctl/flightctl/internal/agent/device/errors"
+	"github.com/flightctl/flightctl/internal/agent/device/fileio"
+	"github.com/flightctl/flightctl/internal/container"
+	"github.com/flightctl/flightctl/pkg/log"
+	"golang.org/x/net/context"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+var _ Manager = (*manager)(nil)
```
Comment (DakCrowder) [inline internal/agent/device/spec/manager.go:22]: Can I get a brief explanation of what this `var _ Manager = (*manager)(nil)` syntax is used for?

**Review by DakCrowder — APPROVED** [review]
:rocket:

File: internal/agent/device/spec/manager.go (hunk 3 above)
Comment (hexfusion) [reply internal/agent/device/spec/manager.go:22]: Hey great question this syntax is an assertion (guarantee) that at compile time the object `*manager` implements the interface `Manager`. It's sort of a cheap safety net.

File: internal/agent/device/errors/errors.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+	// networking
+	ErrNoContent   = errors.New("no content")
+	ErrNilResponse = errors.New("received nil response")
+
+	// authentication
+	ErrAuthenticationFailed = errors.New("authentication failed")
+)
+
+// TODO: tighten up the retryable errors ideally all retryable errors should be explicitly defined
```
[the remark this answers is not in the document]
Comment (hexfusion) [reply internal/agent/device/errors/errors.go:52]: More fun to come here :)

File: internal/agent/device/spec/manager.go (hunk 3 above)
Comment (hexfusion) [reply internal/agent/device/spec/manager.go:22]: To see how it works adjust the signature of one of the methods in the implementation and try to compile.

Comment (DakCrowder) [reply internal/agent/device/spec/manager.go:22]: Ok cool, thanks for the info!!

File: internal/agent/device/device.go [hunk 5]
```diff
@@ -236,118 +289,100 @@ func (a *Agent) beforeUpdateApplications(ctx context.Context, _, desired *v1alph
 				return true, nil
 			})
 			if err != nil {
-				return fmt.Errorf("failed to pull image %q: %w", providerImage, err)
+				return fmt.Errorf("creating enrollment request: %w", err)
```
Comment (celdrake) [inline internal/agent/device/device.go:292]: Nit: Possible rebase error? The previous message was better.

**Comment by siserafin** [comment]
I run some manual tests, the new device error status, when the application image is missing the appType label, works well:
image


and also the device agent error logs:
`level=error msg="Marking template version 3 as failed: before update: applications: no retry: getting application type: required label not found: appType, quay.io/sserafin/flightctl-agent-centos"`
`level=error msg="Reconciliation failed for version 3: before update: applications: no retry: gett"ng application type: required label not found: appType, quay.io/sserafin/flightctl-agent-centos"`

File: internal/agent/device/device.go (hunk 5 above)
Comment (hexfusion) [reply internal/agent/device/device.go:292]: I swear I fixed this haha

Comment (hexfusion) [reply internal/agent/device/device.go]: fixed

**Review by celdrake — APPROVED** [review]

**Pull request merged**